tests/integration/test_prometheus_scrape_multiunit.py

```diff
@@ -0,0 +1,385 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""This test module tests the prometheus_scrape interface with multiple-to-multiple units related.
+
+This test scaling up/down both sides of the relation, and upgrading.
+
+1. Deploy several units of prometheus and several units of a "provider" charm, and relate them.
+2. Confirm all units of prometheus have the correct and same targets and rules.
+3. Upgrade prometheus.
+4. Scale prometheus up and down.
+5. Scale the "provider" charm up and down.
+"""
+
+import asyncio
+import logging
+
+import pytest
+import yaml
+from deepdiff import DeepDiff
+from helpers import (
+    check_prometheus_is_ready,
+    get_prometheus_active_targets,
+    get_prometheus_config,
+    get_prometheus_rules,
+    oci_image,
+    run_promql,
+)
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+prometheus_app_name = "prometheus"
+# prometheus_resources = {"prometheus-image": oci_image("./metadata.yaml", "prometheus-image")}
+prometheus_resources = {"prometheus-image": "prom/prometheus:v2.35.0"}
+tester_app_name = "tester"
+tester_resources = {
+    "prometheus-tester-image": oci_image(
+        "./tests/integration/prometheus-tester/metadata.yaml",
+        "prometheus-tester-image",
+    )
+}
+num_units = 2  # Using the same number of units for both prometheus and the tester
+
+# The period of time required to be idle before `wait_for_idle` returns is set to 90 sec because
+# unit upgrades were observed to take place 40-70 seconds apart.
+idle_period = 90
```

Comment:
I think if you saw upgrades taking 70 seconds, I'd go for something more like 200s. 90s seems to be cutting it kind of close.

Reply:
Increasing `idle_period` would increase the time it takes itests to run.
I think that as long as we don't know how this is set by juju, might as well keep it lower than higher and increase when the need arises.